File: annotations/src/main/java/com/siimkinks/sqlitemagic/annotation/SubmoduleDatabase.java
package com.siimkinks.sqlitemagic.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Define a submodule database configuration.
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS)
public @interface SubmoduleDatabase {
  /**
   * Unique name for classifying submodule database configuration.
   *
   * @return Submodule name
   */
  String value();
}

File: annotations/src/main/kotlin/com/siimkinks/sqlitemagic/annotation/SubmoduleDatabase.kt
package com.siimkinks.sqlitemagic.annotation

import kotlin.annotation.AnnotationRetention.BINARY
import kotlin.annotation.AnnotationTarget.CLASS
import kotlin.reflect.KClass

/**
 * Defines a submodule database configuration.
 *
 * A module may contain one database configuration, defined with either [Database] or
 * [SubmoduleDatabase].
 *
 * @property value Non-empty unique name that identifies the submodule database configuration.
 * @property externalTransformers Classes that contain transformer functions from external modules.
 */
@Target(CLASS)
@Retention(BINARY)
annotation class SubmoduleDatabase(
  val value: String,
  val externalTransformers: Array<KClass<*>> = []
)
